src/disjoint_mut.rs

```diff
@@ -0,0 +1,704 @@
+    where
+        F: FnMut() -> V,
+    {
+        self.inner.get_mut().resize_with(new_len, f)
+    }
+}
+
+unsafe impl<V> AsMutPtr for Vec<V> {
+    type Target = V;
+
+    unsafe fn as_mut_slice(ptr: *mut Self) -> *mut [Self::Target] {
+        // SAFETY: The safety precondition of this method requires that we can
+        // immutably dereference `ptr`.
+        let len = unsafe { (*ptr.cast_const()).len() };
+        // SAFETY: Mutably dereferencing and calling `.as_mut_ptr()` does not
+        // materialize a mutable reference to the underlying slice according to
+        // its documentated behavior, so we can still allow concurrent immutable
+        // references into that underlying slice.
+        let data = unsafe { (*ptr).as_mut_ptr() };
+        ptr::slice_from_raw_parts_mut(data, len)
+    }
+
+    fn len(&self) -> usize {
+        self.len()
+    }
+}
+
+unsafe impl<V, const N: usize> AsMutPtr for [V; N] {
+    type Target = V;
+
+    unsafe fn as_mut_slice(ptr: *mut Self) -> *mut [V] {
+        ptr::slice_from_raw_parts_mut(ptr as *mut V, N)
+    }
+
+    fn len(&self) -> usize {
+        N
+    }
+}
+
+impl<V, const A: usize, const B: usize> DisjointMut<[[V; A]; B]> {
+    // TODO(SJC): Add debug ranges for this method or make this return a raw pointer?
+    #[allow(unused)]
+    pub unsafe fn nd_index_mut(&self, index: [usize; 2]) -> &mut V {
+        assert!(index[0] < B);
+        assert!(index[1] < A);
+        let array = self.inner.get() as *mut [V; A];
+        let array = array.wrapping_add(index[0]) as *mut V;
+        // SAFETY: The preconditions of our `index_mut`  and `nd_index_mut`
+        // safety imply that the indexed region we are mutably borrowing is not
+        // concurrently borrowed and will not be borrowed during the lifetime of
+        // the returned reference. We ensure that the indices are in-bounds with
+        // the two asserts above.
+        unsafe { &mut *array.wrapping_add(index[1]) }
+    }
+}
+
+impl<V: Copy, C: AlignedByteChunk> DisjointMut<AlignedVec<V, C>> {
+    pub fn resize(&mut self, new_len: usize, value: V) {
+        self.inner.get_mut().resize(new_len, value)
+    }
+}
+
+#[cfg(debug_assertions)]
+#[test]
+fn test_pointer_write_debug() {
+    let mut v: DisjointMut<Vec<[u8; 4]>> = Default::default();
+    v.resize(10, [0u8; 4]);
+
+    let guard = v.index(0..);
+    let borrow = &guard[..];
+    let ptr = v.as_mut_ptr().wrapping_offset(3) as *mut u8;
+    unsafe {
+        ptr.wrapping_offset(2).write(42);
+    }
+
+    // Miri complains if we dereference the guard after the pointer write here
+    // instead of dereferencing and borrowing the slice before the write. I
+    // think this is a Miri limitation although to be safe we might want to
+    // change how we store the borrow in the guard. This is not an issue in
+    // release mode without the guards (see test_pointer_write_release below).
+    //
+    // assert_eq!(guard[4][0], 0);
```

Comment:
Should we have `#[should_panic]` tests?

Reply:
Good idea, I'll write some tomorrow.

Reply:
Nope, I take it back. Wrote one and realized that we panic in a destructor which is unrecoverable. It's fine, I don't think we should spend more time here.

Reply:
Ahh. Yeah if tests are tricky it's not worth it.